core/controller/src/main/scala/whisk/core/loadBalancer/SharedDataService.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.loadBalancer
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.{InitialStateAsEvents, MemberEvent, MemberRemoved, MemberUp, UnreachableMember}
+import akka.cluster.ddata.{DistributedData, PNCounterMap, PNCounterMapKey}
+import akka.cluster.ddata.Replicator._
+import whisk.common.AkkaLogging
+import whisk.core.WhiskConfig
+import com.typesafe.config.{Config, ConfigFactory}
+
+case class IncreaseCounter(key: String, value: Long)
+case class DecreaseCounter(key: String, value: Long)
+case class ReadCounter(key: String)
+case class RemoveCounter(key: String)
+case object GetMap
+
+/**
+ * Companion object to specify actor properties from the outside, e.g. name of the shared map
+ */
+object SharedDataService {
+  val requiredProperties = Map(WhiskConfig.controllerSeedNodes -> null)
+
+  def props(storageName: String): Props = Props(new SharedDataService(storageName))
+
+  /**
+   * Add seed nodes if cluster provider is specified, otherwhise return the existing config.
+   * Parse akka seed nodes this way until either of these 2 issues is resolved:
+   * https://github.com/akka/akka/issues/23600
+   * https://github.com/typesafehub/config/issues/69
+   * @return Updated Config
+   */
+  def addAkkaSeedNodesToConf(whiskConf: WhiskConfig): Config = {
+    val conf = ConfigFactory.load()
+
+    val cluster = conf.getString("akka.actor.provider")
+
+    if (cluster == "cluster") {
+      val seedNodes = whiskConf.controllerSeedNodes
```

Comment:
configuring seed nodes statically is very limiting; it would be better to make this extensible, so that various deployment types (kube, mesos) can dictate how "seed nodes" are resolved at runtime, and then init the Akka cluster explicitly AFTER seed nodes are determined (dynamically or from config): 
```
    val seedNodes = SeedConfigSource.getSeedNodes(config)
    Cluster(system).joinSeedNodes(seedNodes.toList)
```
If you make SeedConfigSource either an SPI or a class that allows various approaches, it would enable different deployments the flexibility to register seed nodes as they see fit.

Reply:
👍  We know Mesos and Kube are in the picture and it's not a big stretch to provide support to automatically discover seed nodes using the APIs they provide. Mesos could come later as 1 PR and so does Kube, but it would be good to design the code with this in mind.

Reply:
Not too familiar with both Kube or Mesos APIs but does it make a difference if you determine the seed nodes outside of the component via whichever API and then pass it on startup vs. having the component look them up on startup? Does that break Kube/Mesos deployment models?

Reply:
The idea of this PR was the initial enablement of akka cluster functionality (see feature toggle), we can iterate over it once we put different deployment models into picture. Probably need to take time to familiarize myself with Mesos and Kube.

Reply:
@markusthoemmes @vvraskin the current impl launches the container system with the assumption that the config contains seed nodes, i.e. the actor system initiates clustering automatically via `implicit val actorSystem = ActorSystem("controller-actor-system", SharedDataService.addAkkaSeedNodesToConf(config))`. I'm suggesting that the more flexible approach is to explicitly initiate clustering at some time after the actorsystem is started, using `Cluster(system).joinSeedNodes(seedNodesList)`, which does change the semantics of startup slightly, but allows situations like "actor system and remoting can initialize, before clustering gossip begins", which is good form for doing things like "only allow clustering to proceed if the startup proceeds to the point where remoting can be enabled". Otherwise you are likely to have a noisey (at least) or failure (at worst) during startup in a more dynamic env, or even in a static env, if the seed nodes are not started before the nodes that attempt to join as a cluster (esp in the case where "all the nodes are seed nodes")